WCS: driver: add support to caching XML files and version 2.0.1

ajolma opened the pull request:
https://lists.osgeo.org/pipermail/gdal-dev/2017-November/047733.html

rouault commented on gdal/frmts/wcs/wcsdataset201.cpp:
```diff
@@ -0,0 +1,781 @@
+ *
+ * Project:  WCS Client Driver
+ * Purpose:  Implementation of Dataset class for WCS 2.0.
+ * Author:   Frank Warmerdam, warmerdam@pobox.com
```
you should put yourself as Author there

rouault commented on gdal/frmts/wcs/wcsdataset201.cpp:
```diff
@@ -0,0 +1,781 @@
+    // todo: updatesequence
+
+    std::vector<CPLString> domain = Split(CPLGetXMLValue(psService, "Domain", ""), ",");
+    const char *x = domain[0].c_str();
```
add sanity check that domain.size() == 2 (or >= 2 ?)

rouault commented on gdal/frmts/wcs/wcsdataset201.cpp:
```diff
@@ -0,0 +1,781 @@
+            continue;
+        }
+        std::vector<CPLString> params = Split(FromParenthesis(dimensions[i]), ",");
+        request += "&SUBSET=" + dim + "%28";
```
you could probably use CPLEscapeString(str, -1, CPLES_URL) to do the URL escaping

rouault commented on gdal/frmts/wcs/wcsdataset201.cpp:
```diff
@@ -0,0 +1,781 @@
+            CPLString order = CPLGetXMLValue(axis, "sequenceRule.axisOrder", "");
+            CPLString rule = CPLGetXMLValue(axis, "sequenceRule", "");
+            if (!(order == "+1" && rule == "Linear")) {
+                CPLError(CE_Failure, CPLE_AppDefined, "The grid is not linear and increasing from origo.");
```
origo ?

rouault commented on gdal/frmts/wcs/wcsdataset201.cpp:
```diff
@@ -0,0 +1,781 @@
+        // the offset order should be swapped
+        // Rasdaman does it
+        // MapServer and GeoServer not
+        if (swap_grid_axis) {
```
add sanity check about offset.size()

rouault commented on gdal/frmts/wcs/wcsdataset201.cpp:
```diff
@@ -0,0 +1,781 @@
+/*                                                                      */
+/************************************************************************/
+
+CPLString WCSDataset201::GetSubdataset(CPLString coverage)
```
const CPLString&

rouault commented on gdal/frmts/wcs/wcsdataset201.cpp:
```diff
@@ -0,0 +1,781 @@
+        for (int i = 0; metadata[i] != NULL; ++i) {
+            char *key;
+            CPLString url = CPLParseNameValue(metadata[i], &key);
+            if (strstr(key, "SUBDATASET_") && strstr(key, "_NAME")) {
```
you could add a check for key not being null

rouault commented on gdal/frmts/wcs/wcsdataset201.cpp:
```diff
@@ -0,0 +1,781 @@
+
+    std::vector<CPLString> slow = Split(bbox[0], " ", axis_order_swap);
+    std::vector<CPLString> shigh = Split(bbox[1], " ", axis_order_swap);
+    // todo: if our x,y domain is not the first two? use domain_indexes?
```
sanitfy checks to add on slow.size() and shigh.size() ?

rouault commented on gdal/frmts/wcs/wcsdataset201.cpp:
```diff
@@ -0,0 +1,781 @@
+        swap_grid_axis = !CPLGetXMLBoolean(psService, "NoGridAxisSwap");
+    }
+    path = "limits.GridEnvelope";
+    std::vector<std::vector<int> > size = ParseGridEnvelope(CPLGetXMLNode(grid, path), swap_grid_axis);
```
sanity check on size.size() to add ?

rouault commented on gdal/frmts/wcs/wcsdataset201.cpp:
```diff
@@ -0,0 +1,781 @@
+        // todo: support CRS override, it requires warping the grid to the new CRS
+        CPLError(CE_Failure, CPLE_AppDefined, "CRS override not yet supported.");
+        return false;
+        SetGeometry(grid_size, origin, offsets);
```
dead code

rouault commented on gdal/frmts/wcs/wcsdataset201.cpp:
```diff
@@ -0,0 +1,781 @@
+    // if Range is set in service that may limit the fields
+    int fields = ParseRange(coverage, &metadata);
+    if (fields == 0) {
+        //return false;
```
code to remove, or actually return false ?

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+ *
+ * Project:  WCS Client Driver
+ * Purpose:  Implementation of utilities.
+ * Author:   Frank Warmerdam, warmerdam@pobox.com
```
I'm pretty sure you are the Author of this file and the sole copyright holder

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+    return str2;
+}
+
+CPLString URLRemoveKey(const char *url, CPLString key)
```
const CPLString& key
And use a local variable for key + "="

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+    return retval;
+}
+
+std::vector<CPLString> SwapFirstTwo(std::vector<CPLString> array)
```
Either use
std::vector<CPLString>& SwapFirstTwo(std::vector<CPLString>& array) if you want modify in place and return a reference to the modified object
Or
std::vector<CPLString> SwapFirstTwo(const std::vector<CPLString>& array) if you plan to return a modified copy

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+    return array;
+}
+
+CPLString Join(std::vector<CPLString> array, const char *delim, bool swap_the_first_two)
```
const ... &

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+    return str;
+}
+
+std::vector<int> Ilist(std::vector<CPLString> array,
```
const ... &

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+    return retval;
+}
+
+std::vector<double> Flist(std::vector<CPLString> array,
```
const ... &

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+    return retval;
+}
+
+int IndexOf(CPLString str, std::vector<CPLString> array)
```
const ... &

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+    return index;
+}
+
+int IndexOf(int i, std::vector<int> array)
```
const ... &

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+    return index;
+}
+
+std::vector<int> IndexOf(std::vector<CPLString> strs, std::vector<CPLString> array)
```
const ... &

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+    return retval;
+}
+
+bool Contains(std::vector<int> array, int value)
```
const ... &

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+    return false;
+}
+
+CPLString FromParenthesis(CPLString s)
```
const ... &

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+    return s.substr(beg + 1, end - beg - 1);
+}
+
+std::vector<CPLString> ParseSubset(std::vector<CPLString> subset_array, CPLString dim)
```
const ... &

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+/*      FileIsReadable                                                  */
+/* -------------------------------------------------------------------- */
+
+bool FileIsReadable(CPLString filename)
```
const ... &

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+/*      MakeDir                                                         */
+/* -------------------------------------------------------------------- */
+
+bool MakeDir(CPLString dirname)
```
const ... &

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+        char **folder = VSIReadDir(cache_dir);
+        int size = folder ? CSLCount(folder) : 0;
+        for (int i = 0; i < size; i++) {
+            if (*folder[i] == '.') {
```
folder[i][0] would be more readable

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+/*      DeleteEntryFromCache                                            */
+/* -------------------------------------------------------------------- */
+
+bool DeleteEntryFromCache(CPLString cache_dir, CPLString  key, CPLString value)
```
const ... &

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+        char **folder = VSIReadDir(cache_dir);
+        int size = folder ? CSLCount(folder) : 0;
+        for (int i = 0; i < size; i++) {
+            if (*folder[i] == '.') {
```
folder[i][0]

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+/*      FromCache                                                       */
+/* -------------------------------------------------------------------- */
+
+int FromCache(CPLString cache_dir, CPLString &filename, CPLString url)
```
const CPLString& cache_dir and url

So retval can be 0, 1 or -1 ? Perhaps use an enum instead of int for the return value

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+    return retval;
+}
+
+CPLXMLNode *AddSimpleMetaData(char ***metadata,
```
const CPLString& from. And likely const CPLString& path as well, unless the intent is really to return a modified version to the caller ?

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+    return node2;
+}
+
+CPLString GetKeywords(CPLXMLNode *root,
```
const CPLString &

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+            unsigned int pajazzo = 0, i = 0, a = 0, b = 0;
+            while (1) {
+                unsigned int c = i < epsg_codes.size() ? epsg_codes[i] : 0;
+                if (pajazzo == 1) {
```
pajazzo ?

ajolma replied:
Pajazzo (pajatso) is (was) a popular game machine played with coins in Finland. Here we keep on feeding the loop with EPSG codes (coins) and some are saved and some are lost. Greatly reduces the long list given by GeoServer. And many coins are lost. :) Pajatso is also the Finnish name for an Italian opera Pagliacci.

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+// if appropriate, try to create WKT description from CRS name
+// return false if failure
+// appropriate means, that the name is a real CRS
+bool CRS2Projection(CPLString crs, OGRSpatialReference *sr, char **projection)
```
const CPLString&

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -0,0 +1,698 @@
+    return false;
+}
+
+bool CRSImpliesAxisOrderSwap(CPLString crs, bool &swap, char **projection)
```
const ... &

rouault commented on gdal/frmts/wcs/wcsutils.h:
```diff
@@ -0,0 +1,115 @@
+ *
+ * Project:  WCS Client Driver
+ * Purpose:  Implementation of utilities.
+ * Author:   Frank Warmerdam, warmerdam@pobox.com
```
Same remark as for wcsutils.cpp regarding author & copyright holders

rouault commented on gdal/frmts/wcs/wcsutils.h:
```diff
@@ -0,0 +1,115 @@
+#include "cpl_string.h"
+#include "cpl_minixml.h"
+#include <vector>
+
```
There's a chance that those method names conflict with other C++ symbols in GDAL, or in user code. So I'd suggest to enclods them in a C++  namespace WCSUtils { .... }   and then "using namespce WCSUtils" in the cpp files where you need them

ajolma replied:
done

ajolma commented on gdal/frmts/wcs/frmt_wcs.html:
```diff
@@ -47,9 +47,9 @@
 <li> <b>NoDataValue</b>: The nodata value to use for all the bands (blank for none).  Normally defaulted from CoverageOffering info.
 <li> <b>Timeout</b>: The timeout to use for remote service requests.  If not provided, the libcurl default is used.
-<li> <b>UserPwd</b>: May be supplied with <i>userid:password</i> to pass a
+<li> <b>USERPWD</b>: May be supplied with <i>userid:password</i> to pass a
```
To me it looks that the use of UserPwd in the docs was a typo since it is upper case in the code and it is a HTTP parameter that is upper case everywhere else and the other HTTP parameter below it is upper case.

rouault replied:
I indeed see CPLGetXMLValue( psService, "USERPWD", NULL ) in the code, but looking at CPLGetXMLValue() / CPLGetXMLNode() it uses EQUAL() which is case insensitive. So both will work. We just need to be consistent in the way we at least document this (and in the code as well preferably)

ajolma commented on gdal/frmts/wcs/frmt_wcs.html:
```diff
@@ -60,6 +60,173 @@
 position by default.
 </ul>
 
+<h3>"WCS:" format for dataset name, caching, and version 2.0.1 servers</h3>
```
Ok, I'll rewrite the whole page.

ajolma commented on gdal/frmts/wcs/frmt_wcs.html:
```diff
@@ -60,6 +60,173 @@
+        can't determine the band count and band data type from the
+        capabilities or coverage descriptions. This option may be
+      needed if GetCoverage request fails.)</li>
+    <li><b>OriginNotCenter100</b>=<i>TRUE</i> Some MapServers set the
```
I think it is this
https://github.com/mapserver/mapserver/blob/branch-7-0/mapwcs.c#L1177

rouault replied:
OK, so that's *all* (not some) MapServer WCS 1.0.0 implementations.

ajolma replied:
I guess I was expecting it to be fixed real soon now.

rouault replied:
Regarding WCS 1.0, the standard isn't reallly clear about the convention to be used, whereas WCS 1.1 make it explicit. So I'd prefer not touching WCS 1.0 implementation

ajolma replied:
The current code is based on the assumption that the 'origin' in DescribeCoverage is at the center of pixel. That's in gmlcoverage.cpp and I haven't changed that. In the DescribeCoverage I get from http://194.66.252.155/cgi-bin/BGS_EMODnet_bathymetry/ows?service=WCS&request=DescribeCoverage&version=1.0.0&coverage=BGS_EMODNET_CentralMed-MCol the 'origin' is one on the boundary of the envelope. That leads currently to a half-pixel shift in the envelope gdalinfo reports. In my code I have now the OriginAtBoundary option, which shifts the geotransform origin that half-pixel (in any version). It's an option, thus not applied by default.

ajolma commented on gdal/frmts/wcs/frmt_wcs.html:
```diff
@@ -60,6 +60,173 @@
+    GridOffsets should have four values. However, some servers expect
+    there to be only two (MapServer and ArcGIS) and MapServer expects
+      them to be both positive.
+    <li><b>GridCRS</b>=<i>TRUE</i> If a WCS 1.1 request is not
```
It was not done before. Probably it is not a problem. I'll check.

ajolma replied:
ok, I think I can let it be GridCRSOptional if it was there for a reason.

ajolma commented on gdal/frmts/wcs/frmt_wcs.html:
```diff
@@ -60,6 +60,173 @@
+    <li><b>GridAxisLabelSwap</b>=<i>TRUE</i> When reading the grid
+      axis labels from 2.0.1 coverage description they are swapped
+      even though there is no GridFunction to tell so. May be needed
+      for MapServer in version 2.0.1 if the CRS axis order is
```
I can add that as an exception. The code does not look into axis names. They are just strings to it.

ajolma replied:
It's lat lon<b>g</b> :)

ajolma commented on gdal/frmts/wcs/wcsdataset.cpp:
```diff
@@ -497,6 +101,51 @@ WCSDataset::~WCSDataset()
+    adfGeoTransform[4] = offsets[1].size() == 1 ? 0.0 : offsets[1][0];
+    adfGeoTransform[5] = offsets[1].size() == 1 ? offsets[1][0] : offsets[1][1];
+    
+    adfGeoTransform[0] -= adfGeoTransform[1] * 0.5;
```
wcsdataset100.cpp does not call this method. Maybe it should for clarity's sake but I did not want to change the existing code too much.

rouault replied:
Fine, perhaps just add a comment that this isn't used by WCS 1.0.0

ajolma commented on gdal/frmts/wcs/wcsdataset.cpp:
```diff
@@ -2049,29 +801,424 @@ int WCSDataset::Identify( GDALOpenInfo * poOpenInfo )
+        CPLHTTPDestroyResult(psResult);
+        return NULL;
+    }
+    capabilities = capabilities->psNext; // from ?XML to root
```
The modification is removing the namespace. I don't think there are any other modifications. Or maybe I'm missing the point?

rouault replied:
I'm speaking about the following pattern
```
CPLXMLNode* capabilities = CPLParseXML....()
capabilities = capabilities->psNext;
```
By doing so, there are several issues:
* first you assume that the first node is a <?xml ....> node. But this is not mandatory in XML
* there could be extra <? ... > or <!-- --> nodes
* you loose the reference to the original pointer returned by CPLParseXML...(), so you will get memory leaks if you just freee the modified capabilities

rouault commented on gdal/frmts/wcs/wcsdataset.cpp:
```diff
@@ -850,31 +851,43 @@ WCSDataset *WCSDataset::CreateFromCapabilities(GDALOpenInfo * poOpenInfo, CPLStr
+    // to avoid hardcoding the name of the Capabilities element
+    // we assume it is the next of the possible <?xml> element
+    CPLXMLNode *capabilities = doc.get();
+    if (EQUAL(capabilities->pszValue, "?xml")) {
```
I'd suggest rewriting this as

```
while( capabilities != NULL && (capabilities->eType != CXT_Element || capabilities->pszValue[0] == '?') )
{
     capabilities = capabilities->psNext;
}
```

rouault commented on gdal/frmts/wcs/wcsdataset201.cpp:
```diff
@@ -664,6 +664,11 @@ bool WCSDataset201::ExtractGridInfo()
     bool swap_grid_axis_labels = swap_grid_axis || CPLGetXMLBoolean(psService, "GridAxisLabelSwap");
     std::vector<CPLString> grid_axes = Split(CPLGetXMLValue(grid, path, ""), " ", swap_grid_axis_labels);
+    // autocorrect MapServer thing
+    if (grid_axes[0] == "lat" && grid_axes[0] == "long") {
```
typo on the second index should be grid_axes[1] == "long". And shouldn't there be a grid_axes.size() >= 2 test ?

ajolma commented on gdal/frmts/wcs/wcsdataset110.cpp:
```diff
@@ -0,0 +1,880 @@
+        bbox_3 = extent[3]; // max Y
+
+    if (axis_order_swap) {
+        bbox_0 = extent[1], // min Y
```
Javascript influence.

ajolma commented:
Lots of those comments - at least the "make this const&" - could perhaps be automatized with some tool? The XML 'get document node', should be, IMO, a method in the XML library.

rouault commented:
> Lots of those comments - at least the "make this const&" - could perhaps be automatized with some tool?

Perhaps. I'm not aware of one

> The XML 'get document node', should be, IMO, a method in the XML library.

Feel free to extend cpl_minixml

ajolma commented on gdal/frmts/wcs/frmt_wcs.html:
```diff
@@ -60,6 +60,173 @@
+      file unless asked to refresh or recreate.
+    </li>
+    <li>
+      Once the coverage offering of the server is known, a coverage
```
Explained now. Capabilities is first fetched.

ajolma commented:
While looking into the pixel center / boundary thing in MapServer I hit another issue with bands (MapServer doesn't seem to honor the request which leads to GDAL bailing out). That came out when gdal_translate was using the driver in a way my earlier tests were not - gdal_translate with just -srcwin. I'll also write new tests (into the Perl test) which tests the boundaries of received data agains those announced in DescribeCoverage. That'll take some time.

ajolma commented:
I've completed the non-scaled gdal_translate based tests. I added BandIdentifier option to force the driver to ignore band specific requests (gdal_translate seems to read the dataset band by band) which caused an error since MapServer does not seem to honor range subsetting in 1.0 or 1.1. I also changed the OriginNotCenter100 to OriginAtBoundary which is indeed needed in 2.0.1 with MapServer. These tests with live servers are done with the Perl program responses/requests.pl, which does the tests and also downloads the server responses and records the URLs the driver makes to be used by the autonomous python test.

rouault commented on gdal/frmts/wcs/wcsutils.cpp:
```diff
@@ -403,52 +418,95 @@ bool DeleteEntryFromCache(const CPLString &cache_dir, const CPLString &key, cons
+        return CE_Failure;
+    }
+        
+    // create a new file into the cache using filename as template
```
instead of using randomness, why not using a hash on the URL ? See https://github.com/OSGeo/gdal/blob/trunk/gdal/ogr/ogrsf_frmts/gmlas/ogrgmlasxsdcache.cpp#L124

ajolma commented:
Note that I had to rewrite the bootstrap code in Open the Nth time since now it's three-level system: Capabilities, Template-Coverage, Specialized-Coverage.

ajolma commented on gdal/frmts/wcs/wcsdataset.h:
```diff
@@ -0,0 +1,220 @@
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
```
done

ajolma commented:
I'm quite happy with this now. I'd like to add two cache methods still. "WCS:" could return a metadataset with all datasets in the cache as subdatasets. And -oo "DELETE_FROM_CACHE=i" could be used to delete entry i from the cache - this is conceptually a bit funny though but something like this is needed. To the python tests i could add tests of URLs generated for non-scaled requests and range and dimension subsetting requests - these are in the live Perl tests (data/wcs/requests.pl).

ajolma commented:
The bug that showed itself in Windows is fixed. The non-scaled request tests (get srcwin 0 0 2 2) are problematic since the response files are quite big (~1,5 MB each).

ajolma commented:
Even, this is getting into a quite good condition. Do you have suggestions what is still needed for merging this with trunk? I believe I have addressed all your comments.

rouault commented on autotest/gdrivers/wcs.py:
```diff
@@ -272,6 +281,269 @@ def old_wcs_5():
+    driver = gdal.GetDriverByName('WCS')
+    if driver is None:
+	    return 'skip'
+    print("Generating various URLs from the driver and comparing them to ones")
```
We generally avoid tests to emit messages. Perhaps just turn that into a comment

rouault commented on autotest/gdrivers/wcs.py:
```diff
@@ -272,6 +281,269 @@ def old_wcs_5():
+            for i, c in enumerate(projwin):
+                projwin[i] = int(c)
+            options = [cache]
+            gdal.Translate("tmp/"+server+version+".tiff", ds, projWin = projwin, width = size, options = options)
```
That temporary file should be deleted

rouault commented on autotest/gdrivers/wcs.py:
```diff
@@ -272,6 +281,269 @@ def old_wcs_5():
+                    print("OpenEx failed: WCS:" + url + "/?" + query)
+                    break
+                options = [cache]
+                gdal.Translate("tmp/"+server+version+".tiff", ds, srcWin = [0,0,2,2], options = options)
```
Same here

ajolma commented:
What do you think about the test files? The non-scaled tests (tiff responses to translates) are pretty big put together. What about downloading them from someplace for tests?

rouault commented:
> What do you think about the test files? The non-scaled tests (tiff responses to translates) are pretty big put together (and I have included so far only one). What about downloading them from someplace for tests?

Ah I assumed the files where small. Cannot you create dummy versions of them with "gdal_translate in.tif out.tif -scale 0 255 0 0 -co SPARSE_OK=YES" ? This should normally create a sparse blank file with the same georeferencing

rouault commented:
Merged in trunk. Closing
